File: packages/go/luna/resources/payments/yoco.go
// Package payments provides South African payment gateway integrations.
package payments

import (
	"context"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"time"

	lunahttp "github.com/eclipse-softworks/luna-sdk-go/luna/http"
)

// Yoco provides Yoco online payment integration.
type Yoco struct {
	client *lunahttp.Client
	config YocoConfig
}

// NewYoco creates a new Yoco instance.
func NewYoco(client *lunahttp.Client, config YocoConfig) *Yoco {
	return &Yoco{
		client: client,
		config: config,
	}
}

// CreatePayment creates a checkout session and returns the redirect URL.
func (y *Yoco) CreatePayment(ctx context.Context, req YocoPaymentRequest) (*YocoPayment, error) {
	currency := req.Currency
	if currency == "" {
		currency = "ZAR"
	}

	body := map[string]interface{}{
		"amount":     req.Amount,
		"currency":   currency,
		"metadata":   req.Metadata,
		"successUrl": req.SuccessURL,
		"cancelUrl":  req.CancelURL,
		"failureUrl": req.FailureURL,
	}

	resp, err := y.client.Request(ctx, lunahttp.RequestConfig{
		Method: "POST",
		Path:   "/v1/payments/yoco/checkouts",
		Body:   body,
	})
	if err != nil {
		return nil, err
	}

	var result struct {
		ID          string                 `json:"id"`
		CheckoutID  string                 `json:"checkoutId"`
		Amount      float64                `json:"amount"`
		Currency    string                 `json:"currency"`
		Status      PaymentStatus          `json:"status"`
		Reference   string                 `json:"reference"`
		RedirectURL string                 `json:"redirectUrl"`
		Metadata    map[string]interface{} `json:"metadata"`
		CreatedAt   time.Time              `json:"createdAt"`
		UpdatedAt   time.Time              `json:"updatedAt"`
	}

	if err := json.Unmarshal(resp.Data, &result); err != nil {
		return nil, err
	}

	return &YocoPayment{
		ID:         result.ID,
		Provider:   "yoco",
		CheckoutID: result.CheckoutID,
		Amount: Amount{
			Value:    int(result.Amount),
			Currency: result.Currency,
		},
		Status:      result.Status,
		Reference:   result.Reference,
		RedirectURL: result.RedirectURL,
		Metadata:    result.Metadata,
		CreatedAt:   result.CreatedAt,
		UpdatedAt:   result.UpdatedAt,
	}, nil
}

// VerifyWebhook verifies the webhook signature.
func (y *Yoco) VerifyWebhook(payload string, signature string) bool {
	mac := hmac.New(sha256.New, []byte(y.config.SecretKey))
	mac.Write([]byte(payload))
	expectedSignature := hex.EncodeToString(mac.Sum(nil))
	return signature == expectedSignature
}

// ProcessWebhook processes a webhook event.
func (y *Yoco) ProcessWebhook(payload map[string]interface{}) *YocoPayment {
	statusMap := map[string]PaymentStatus{
		"payment.succeeded": StatusCompleted,
		"payment.failed":    StatusFailed,
		"payment.cancelled": StatusCancelled,
	}

	eventType, _ := payload["type"].(string)
	paymentData, _ := payload["payload"].(map[string]interface{})

	id, _ := paymentData["id"].(string)
	amount, _ := paymentData["amount"].(float64)
	currency, _ := paymentData["currency"].(string)
	if currency == "" {
		currency = "ZAR"
	}

	status := StatusPending
	if s, ok := statusMap[eventType]; ok {
		status = s
	}

	var metadata map[string]interface{}
	if m, ok := paymentData["metadata"].(map[string]interface{}); ok {
		metadata = m
	}

	return &YocoPayment{
		ID:         fmt.Sprintf("yc_%s", id),
		Provider:   "yoco",
		CheckoutID: id,
		Amount: Amount{
			Value:    int(amount),
			Currency: currency,
		},
		Status:    status,
		Reference: id,
		Metadata:  metadata,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}
}

// Refund requests a refund.
func (y *Yoco) Refund(ctx context.Context, req RefundRequest) (*Refund, error) {
	refundID := fmt.Sprintf("ref_%d", time.Now().UnixMilli())

	amount := 0
	if req.Amount != nil {
		amount = *req.Amount
	}

	return &Refund{
		ID:        refundID,
		PaymentID: req.PaymentID,
		Amount: Amount{
			Value:    amount,
			Currency: "ZAR",
		},
		Status:    "pending",
		Reason:    req.Reason,
		CreatedAt: time.Now(),
	}, nil
}

File: packages/python/luna/resources/payments/yoco.py
"""Luna SDK - Yoco Integration for Python."""

from __future__ import annotations

import hashlib
import hmac
from datetime import datetime
from typing import Any

from luna.http import HttpClient
from .types import (
    YocoConfig,
    YocoPaymentRequest,
    YocoPayment,
    RefundRequest,
    Refund,
    Amount,
)


class Yoco:
    """Yoco online payment integration."""

    def __init__(self, client: HttpClient, config: YocoConfig) -> None:
        self._client = client
        self._config = config

    async def create_payment(self, request: YocoPaymentRequest) -> YocoPayment:
        """Create a checkout session and get redirect URL."""
        response = await self._client.request(
            method="POST",
            path="/v1/payments/yoco/checkouts",  # Proxy through Luna API
            body={
                "amount": request.amount,
                "currency": request.currency,
                "metadata": request.metadata,
                "successUrl": self._config.success_url if hasattr(self._config, "success_url") else None,
                "cancelUrl": self._config.cancel_url if hasattr(self._config, "cancel_url") else None,
                "failureUrl": self._config.failure_url if hasattr(self._config, "failure_url") else None,
            },
        )
        data = response.data
        return YocoPayment(
            id=data["id"],
            provider="yoco",
            checkout_id=data["checkoutId"],
            amount=Amount(value=data["amount"], currency=data["currency"]),
            status=data["status"],
            reference=data["reference"],
            redirect_url=data["redirectUrl"],
            metadata=data.get("metadata"),
            created_at=data["createdAt"],
            updated_at=data["updatedAt"],
        )

    def verify_webhook(self, payload: str, signature: str) -> bool:
        """Verify webhook signature."""
        expected_signature = hmac.new(
            self._config.secret_key.encode(),
            payload.encode(),
            hashlib.sha256,
        ).hexdigest()
        return signature == expected_signature

    def process_webhook(self, payload: dict[str, Any]) -> YocoPayment:
        """Process webhook event."""
        status_map = {
            "payment.succeeded": "completed",
            "payment.failed": "failed",
            "payment.cancelled": "cancelled",
        }

        event_type = payload.get("type", "")
        payment_data = payload.get("payload", {})

        return YocoPayment(
            id=f"yc_{payment_data.get('id', '')}",
            provider="yoco",
            checkout_id=payment_data.get("id", ""),
            amount=Amount(
                value=payment_data.get("amount", 0),
                currency=payment_data.get("currency", "ZAR"),
            ),
            status=status_map.get(event_type, "pending"),
            reference=payment_data.get("id"),
            metadata=payment_data.get("metadata"),
            redirect_url="",
            created_at=datetime.now().isoformat(),
            updated_at=datetime.now().isoformat(),
        )

    async def refund(self, request: RefundRequest) -> Refund:
        """Request a refund."""
        refund_id = f"ref_{int(datetime.now().timestamp() * 1000)}"

        return Refund(
            id=refund_id,
            payment_id=request.payment_id,
            amount=Amount(value=request.amount or 0, currency="ZAR"),
            status="pending",
            reason=request.reason,
            created_at=datetime.now().isoformat(),
        )
